add resilient topo server caching for the full srv keyspace object

demmer opened the pull request:
## What
Add support to the ResilientServer to cache the whole SrvKeyspace object for the configured TTL and a configurable refresh interval in addition to the TTL.

## Why
This change fixes errors that we have observed where vtgate fails to route queries during intermittent consul cluster issues. 

## Details
For entries we don't watch (like the list of Keyspaces), we refresh the cached list from the topo after srv_topo_cache_refresh elapses. If the fetch fails, we hold onto the cached value until srv_topo_cache_ttl elapses.

For entries we watch (like the SrvKeyspace for a given cell), if setting the watch fails, we will use the last known value until srv_topo_cache_ttl elapses and we only try to re-establish the watch once every srv_topo_cache_refresh interval.

I also reused some of the metrics that were being counted for the GetSrvKeyspaceNames handler for the GetSrvKeyspace watch so we can see what is going on under the hood better.

## Testing
In addition to the new unit tests added in this PR, I tested these changes on one of our dev vtgate hosts on which I stopped the consul agent causing all topo calls to fail, and confirmed that queries could be routed for the duration of the cache TTL, but once the TTL elapsed the queries started failing as expected.

alainjobart commented on go/vt/srvtopo/resilient_server_test.go:
```diff
@@ -150,6 +213,19 @@ func TestSrvKeyspaceCachedError(t *testing.T) {
 	}
 }
 
+// TestSrvKeyspaceWatchError will test we properly return the cached topo
```
Probably extra code here that needs to go?

demmer commented:
@alainjobart I finished up the future thoughts so this ready for review.

Note that I ended up making a bunch of changes to try to make the test comprehensive and reliable without slowing down too much, which a challenge since many of these behaviors are dependent on real clocks...

To be safe I marked it as flaky to try to appease Travis but this might end up needing more work if it is too unreliable.

alainjobart commented on go/vt/srvtopo/resilient_server.go:
```diff
@@ -259,46 +289,81 @@ func (server *ResilientServer) GetSrvKeyspace(ctx context.Context, cell, keyspac
 				log.Errorf("%v", err)
+				server.counts.Add(errorCategory, 1)
 				entry.mutex.Lock()
+				if err == topo.ErrNoNode {
```
if c.Err == topo.ErrNoNode  here?

demmer replied:
Ah yes good catch thanks.
Fixed but I should probably add a test for this.

alainjobart commented on go/vt/srvtopo/resilient_server.go:
```diff
@@ -259,46 +289,81 @@ func (server *ResilientServer) GetSrvKeyspace(ctx context.Context, cell, keyspac
 				entry.lastError = err
 				entry.lastErrorCtx = nil
 				entry.mutex.Unlock()
+				cancel()
```
maybe:
defer cancel()
right at the beginning of the go func() would be clearer?

demmer replied:
good idea - fixed.

alainjobart commented on go/vt/srvtopo/resilient_server.go:
```diff
@@ -308,7 +373,7 @@ func (server *ResilientServer) GetSrvKeyspace(ctx context.Context, cell, keyspac
 }
 
-var watchSrvVSchemaSleepTime = 5 * time.Second
+var watchSrvVSchemaSleepTime = sync2.NewAtomicDuration(5 * time.Second)
```
You made that change because the test was giving you a warning in the unit_race?
If it's just for the test, I would change the line that sets the value to be much earlier in the test, before it's used, so you don't need this.

demmer replied:
Yeah exactly -- for some reason I just tried reverting that change to reproduce it and it doesn't happen any more. Yay races.

Regardless your idea is way better so I just did that.

alainjobart commented on go/vt/srvtopo/resilient_server.go:
```diff
@@ -28,14 +28,16 @@ import (
 var (
-	srvTopoCacheTTL = flag.Duration("srv_topo_cache_ttl", 1*time.Second, "how long to use cached entries for topology")
+	srvTopoCacheTTL     = flag.Duration("srv_topo_cache_ttl", 1*time.Second, "how long to use cached entries for topology")
+	srvTopoCacheRefresh = flag.Duration("srv_topo_cache_refresh", 1*time.Second, "how frequently to refresh the topology for cached entries")
```
I wonder where to add this, but it would be nice to have scenarios that describe the behavior, something like:

* for entries we don't watch (like the list of Keyspaces), we refresh the list only after srv_topo_cache_ttl.
* for entries we watch, and setting the watch fails, we will use the last know value for srv_topo_cache_ttl.

demmer replied:
Good idea. I added a block comment describing the behavior in the code itself. I can try to edit the flag description to be more clear but I would worry that would be too long to be useful.

demmer commented:
@alainjobart This is now passing tests and I think should be ready to go.

demmer commented:
@alainjobart As per conversation in Slack I've cleaned up the commit history and rebased to remove the various back and forth trying to get the tests to pass.

The remaining commits seem logical enough to me to keep them as-is.

alainjobart commented:
LGTM
